### ISSUE TITLE
Support anyOf in response

### ISSUE THREAD

**Issue by ferdikoomen** [ferdikoomen/openapi-typescript-codegen#141]
Right now we don't support anyOf in the response (i think):

```
"responses": {
	"200": {
		"description": "The request was succesfull",
		"schema": {
		"anyOf": [
			{ "$ref": "#/definitions/Dog" },
			{ "$ref": "#/definitions/Cat" },
			{ "$ref": "#/definitions/Animal" }
		]
	}
}
```

**Comment by kcappieg** [comment]
It seems "anyOf" is not supported in any model. A couple code files are depending on the type definition to be a string, not an array of strings. (First error was thrown to me in file `...dist/openApi/v3/parser/stripNamespace.js`. I monkey-patched the first error to see if I could get the whole thing working, but then got another error somewhere further down the chain.)

Seems like a useful project, hope you keep working on it.

**Comment by ferdikoomen** [comment]
Hey @kcappieg Would it be possible to send me your spec file (in an email or dm). I can have a quick look to see what needs to happen. Im always happy to support the full OpenAPI spec.

**Comment by ferdikoomen** [comment]
i have some use cases, but it's always great to have some real-world test data.

**Comment by kcappieg** [comment]
Turns out I misunderstood something about the openapi spec. I'm new to OpenAPI and using Spotlight as a GUI editor. When building models, it allows you to specify multiple types that it outputs as an array in the JSON OpenAPI spec. This seems to be the same as `anyOf`, but a different syntax. Reviewing the docs from swagger.io... this does not actually seem like it's valid syntax :(

So instead of

```JSON
"myModel": {
  "anyOf": [/*some types*/]
}
```

It outputs:

```JSON
"myModel": {
  "type": ["string","integer"]
}
```

The dangers of trusting a GUI to do the heavy lifting for you...

Thanks for the response, but it seems I need to better understand the spec before I can complain something's not working!

**Comment by artem-tim** [comment]
Hi, hopping onto the issue - we also considering to use this (instead of that awful java codegen) for our project, but we would need the support for anyOf/oneOf as well
The spec is here https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/
Basically instead of only being able to have a single schema :
```
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Animal'
```
We would have a possibility to allow several possible inputs for the API :
```
      requestBody:
        content:
          application/json:
            schema:
              oneOf:
                - $ref: '#/components/schemas/Cat'
                - $ref: '#/components/schemas/Dog'
```

However the spec allows oneOf etc at multiple levels, even within the schema itself (it should be possible to replace 1 object by an oneOf array of sub-objects)

Do you think it could be added ? I don't mind contributing if there is not enough bandwidth

**Comment by ferdikoomen** [comment]
Hi @artem-tim i have some other changes pending as well, like the support for response header parsing. Let me check if i can fit this in. Might have some time over the weekend to update this. Thanks for the info, this should help a lot.

**Comment by ferdikoomen** [comment]
@artem-tim I investigated the issue and started to make some changes. I need to call it a day now. Will ping you when there is a working branch.

**Comment by patrikniebur** [comment]
+1 on this. Great project, thank you for doing this. Would like to use it in our workflow too but need support for allOf and oneOf

**Comment by thetrime** [comment]
I've noticed that https://github.com/diverta/kuroco-sdk/blob/master/src/generator/openApi/v3/parser/getModel.ts implements this, though it's not completely clear that this was the first project to do so. Patching it in to this repo gives me the results I'd expect to see (though I had to also add es2019 to the lib section of tsconfig.json so I could get things like Array.flat). I'll see if I can track down the original implementor...

**Comment by thetrime** [comment]
It looks like diverta really did come up with that change first. Are you interested in a PR if I tidy it up to keep it ES2017?

**Comment by ferdikoomen** [comment]
@thetrime Sure! sounds like a great idea!

**Comment by thetrime** [comment]
Ok, see https://github.com/ferdikoomen/openapi-typescript-codegen/pull/399

**Comment by ferdikoomen** [comment]
Pushed 0.7.0-beta to NPM, thanks everybody for the input and specially @budde377 for the PR. Lets test the changes and see if they are working well.

**Comment by ferdikoomen** [comment]
Fixed a minor issue, will release the 0.7.0 version today.
